### PR TITLE
Added check if project name starts with number when creating new proj…

### LINKED ISSUE #1203: Forbid creating projects with names starting with number
Currently, I can create a project with name starting with number, but cannot run it:
```
d:\>tns create 123
Project 123 was successfully created

d:>cd 123
d:\123>tns run adnroid
Project name must not begin with a number
```
Better fail on the create step.

### PR DESCRIPTION
**Pull request by TsvetanMilanov**
…ect.

Created project name service with method to ensure the project name does not start with number or if it starts with number the client is informed about the consequences when building for Android.
When invalid project name is entered the client gets warning about the problem and a prompt with choice to create the project with invalid name or to enter valid name.
Added integration tests for the prompts logic.

This fixes #1203

### CODE REVIEW

File: lib/services/project-name-service.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+		private $prompter: IPrompter) { }
+
+	private checkIfNameStartsWithLetter(projectName: string): boolean {
+		let startsWithLetterExpression = /^[a-zA-Z]/;
+		return startsWithLetterExpression.test(projectName);
+	}
+
+	ensureValidName(projectName: string, validateOptions?: {force: boolean}): IFuture<void> {
+		return (() => {
+			if (validateOptions.force) {
```
Comment (Fatme) [inline lib/services/project-name-service.ts]: Should be ` if (validateOptions && validateOptions.force) { `

File: lib/services/project-name-service.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+				return;
+			}
+
+			let projectNameStartsWithLetter = this.checkIfNameStartsWithLetter(projectName);
```
Comment (rosen-vladimirov) [inline lib/services/project-name-service.ts]: you do not need this variable, just call:
```TypeScript
if(this.checkIfNameStartsWithLetter(projectName)) {
```

File: lib/declarations.ts [hunk 3]
```diff
@@ -255,3 +255,16 @@ interface IXmlValidator {
 	 */
 	getXmlFileErrors(sourceFile: string): IFuture<string>;
 }
+
+/**
+ * Describes methods for project name.
+ */
+interface IProjectNameService {
+	/**
+	 * Ensures the passed project name is valid. If the project name is not valida prompts for actions.
```
Comment (rosen-vladimirov) [inline lib/declarations.ts]: `valida` -> `valid`

File: lib/services/project-name-service.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+			let newProjectName = this.$prompter.getString("Enter the new project name:").wait();
+			return this.ensureValidName(newProjectName, validateOptions).wait();
+		}).future<string>()();
+	}
+
+	private promptForForceNameConfirm(): IFuture<boolean> {
+		return (() => {
+			this.$logger.warn("The project name does not start with letter and will fail to build for Android.");
+
+			return this.$prompter.confirm("Do You want to create the project with this name?").wait();
```
Comment (rosen-vladimirov) [inline lib/services/project-name-service.ts]: `You` -> `you`

File: test/project-name-service.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
+/// <reference path=".d.ts" />
+"use strict";
+
+import * as yok from "../lib/common/yok";
```
Comment (rosen-vladimirov) [inline test/project-name-service.ts]:
```TypeScript
import { Yok } from "../lib/common/yok";
```

File: test/project-name-service.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import { ProjectNameService } from "../lib/services/project-name-service";
+import {assert} from "chai";
+import {ErrorsStub, LoggerStub} from "./stubs";
+
+let mockProjectNameValidator = {
+	validate: () => { return true; }
```
Comment (rosen-vladimirov) [inline test/project-name-service.ts]:
```TypeScript
validate: () => true
```

File: test/project-name-service.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+function createTestInjector(): IInjector {
+	let testInjector: IInjector;
+
+	testInjector = new yok.Yok();
+	testInjector.register("projectNameService", ProjectNameService);
+	testInjector.register("projectNameValidator", mockProjectNameValidator);
+	testInjector.register("errors", ErrorsStub);
+	testInjector.register("logger", LoggerStub);
+	testInjector.register("prompter", {
+		confirm: (message: string): IFuture<boolean> => {
```
Comment (rosen-vladimirov) [inline test/project-name-service.ts]:
```TypeScript
confirm: (message: string): IFuture<boolean> =>  Future.fromResult(true), 
getString: (message: string): IFuture<string> => Future.fromResult(dummyString)
```

File: test/project-name-service.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+	return testInjector;
+}
+
+describe("Project Name Service Tests", () => {
+	let testInjector: IInjector;
+	let projectNameService: IProjectNameService;
+	let validProjectName = "valid";
+	let invalidProjectName = "1invalid";
+
+	before(() => {
```
Comment (rosen-vladimirov) [inline test/project-name-service.ts]: before is called before the first test only. We prefer using `beforeEach`, so each test will be on clean setup

File: test/project-name-service.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+	it("returns correct name when valid name is entered", () => {
+		let actualProjectName = projectNameService.ensureValidName(validProjectName).wait();
+
+		assert.deepEqual(actualProjectName, validProjectName);
+	});
+
+	it("returns correct name when invalid name is entered several times and then valid name is entered", () => {
+		let prompter = testInjector.resolve("prompter");
+		prompter.confirm = (message: string): IFuture<boolean> => {
```
Comment (rosen-vladimirov) [inline test/project-name-service.ts]:
```TypeScript
prompter.confirm = (message: string) =>  Future.fromResult(false);
```

File: test/project-name-service.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+		let incorrectInputsCount = 0;
+
+		prompter.getString = (message: string): IFuture<string> => {
+			return (() => {
+				if (incorrectInputsCount < incorrectInputsLimit) {
+					incorrectInputsCount++;
+
+					return invalidProjectName;
+				}
+				else {
```
Comment (rosen-vladimirov) [inline test/project-name-service.ts]: `else` statement must be on the line above

File: test/project-name-service.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+				}
+			}).future<string>()();
+		};
+
+		let actualProjectName = projectNameService.ensureValidName(invalidProjectName).wait();
+
+		assert.deepEqual(actualProjectName, validProjectName);
+	});
+
+	it("returns correct name when invalid name is entered and --force flag is present", () => {
```
Comment (rosen-vladimirov) [inline test/project-name-service.ts]: returns the invalid name when...
Sounds better for name of the test

File: test/project-name-service.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+	it("returns correct name when invalid name is entered several times and then valid name is entered", () => {
+		let prompter = testInjector.resolve("prompter");
+		prompter.confirm = (message: string): IFuture<boolean> => {
+			return (() => {
+				return false;
+			}).future<boolean>()();
+		};
+
+		let incorrectInputsLimit = 20;
```
Comment (rosen-vladimirov) [inline test/project-name-service.ts]: set it to 5 please

File: test/project-service.ts [hunk 13]
```diff
@@ -136,6 +142,19 @@ class ProjectIntegrationTest {
 
 		this.testInjector.register("options", Options);
 		this.testInjector.register("hostInfo", HostInfo);
+		this.testInjector.register("prompter", {
+			confirm: (message: string): IFuture<boolean> => {
```
Comment (rosen-vladimirov) [inline test/project-service.ts]:
```TypeScript
confirm: (message: string): IFuture<boolean> => Future.fromResult(true)
```

File: test/project-service.ts [hunk 14]
```diff
@@ -273,6 +292,153 @@ describe("Project Service Tests", () => {
...
+				let projectIntegrationTest = new ProjectIntegrationTest();
+				let tempFolder = temp.mkdirSync("project");
+				let projectName = invalidProjectName;
+				helpers.isInteractive = () => { return true; };
+
+				let options: IOptions = projectIntegrationTest.testInjector.resolve("options");
+				options.force = true;
+
+				options.path = tempFolder;
+				options.copyFrom = projectIntegrationTest.getNpmPackagePath("tns-template-hello-world").wait();
```
Comment (rosen-vladimirov) [inline test/project-service.ts]: do we need this line?

File: test/project-service.ts [hunk 15]
```diff
@@ -273,6 +292,153 @@ describe("Project Service Tests", () => {
...
+
+			afterEach(() => {
+				helpers.isInteractive = originalIsInteractive;
+			});
+
+			it("creates project when is interactive and incorrect name is specified and the --force option is set", () => {
+				let projectIntegrationTest = new ProjectIntegrationTest();
+				let tempFolder = temp.mkdirSync("project");
+				let projectName = invalidProjectName;
+				helpers.isInteractive = () => { return true; };
```
Comment (rosen-vladimirov) [inline test/project-service.ts]:
```TypeScript
helpers.isInteractive = () => true;
```
Also you can move this line in the beforeEach statement. You are using it in all of your tests (you will have to keep it only in the last test, where the passed value is `false`

File: test/project-service.ts [hunk 16]
```diff
@@ -273,6 +292,153 @@ describe("Project Service Tests", () => {
...
+			beforeEach(() => {
+				hasPromptedForString = false;
+			})
+
+			afterEach(() => {
+				helpers.isInteractive = originalIsInteractive;
+			});
+
+			it("creates project when is interactive and incorrect name is specified and the --force option is set", () => {
+				let projectIntegrationTest = new ProjectIntegrationTest();
```
Comment (rosen-vladimirov) [inline test/project-service.ts]: you can move this line in the `beforeEach` statement

File: test/project-service.ts [hunk 17]
```diff
@@ -273,6 +292,153 @@ describe("Project Service Tests", () => {
...
+				let tempFolder = temp.mkdirSync("project");
```
Comment (rosen-vladimirov) [inline test/project-service.ts]: you can move this line in the `beforeEach` statement

File: test/project-service.ts [hunk 18]
```diff
@@ -273,6 +292,153 @@ describe("Project Service Tests", () => {
...
+				let projectName = invalidProjectName;
+				helpers.isInteractive = () => { return true; };
+
+				let options: IOptions = projectIntegrationTest.testInjector.resolve("options");
```
Comment (rosen-vladimirov) [inline test/project-service.ts]: you can move this line in the `beforeEach` statement

File: test/project-service.ts [hunk 19]
```diff
@@ -273,6 +292,153 @@ describe("Project Service Tests", () => {
...
+				projectIntegrationTest.assertProject(tempFolder, projectName, `org.nativescript.${projectName}`).wait();
+			});
+
+			it("creates project when is interactive and incorrect name is specified and the user confirms to use the incorrect name", () => {
+				let projectIntegrationTest = new ProjectIntegrationTest();
+				let tempFolder = temp.mkdirSync("project");
+				let projectName = invalidProjectName;
+				helpers.isInteractive = () => { return true; };
+
+				let prompter = projectIntegrationTest.testInjector.resolve("prompter");
```
Comment (rosen-vladimirov) [inline test/project-service.ts]: you can move this line in the `beforeEach` statement

File: test/project-service.ts [hunk 20]
```diff
@@ -273,6 +292,153 @@ describe("Project Service Tests", () => {
...
+				prompter.confirm = (message: string): IFuture<boolean> => {
```
Comment (rosen-vladimirov) [inline test/project-service.ts]:
```TypeScript
prompter.confirm = (message: string): IFuture<boolean> => true;
```

File: test/project-service.ts [hunk 21]
```diff
@@ -273,6 +292,153 @@ describe("Project Service Tests", () => {
...
+					return (() => {
+						return true;
+					}).future<boolean>()();
+				};
+
+				let options: IOptions = projectIntegrationTest.testInjector.resolve("options");
+
+				options.path = tempFolder;
+				options.copyFrom = projectIntegrationTest.getNpmPackagePath("tns-template-hello-world").wait();
```
Comment (rosen-vladimirov) [inline test/project-service.ts]: same as above, I think we do not need this line

File: test/project-service.ts [hunk 22]
```diff
@@ -273,6 +292,153 @@ describe("Project Service Tests", () => {
...
 			projectIntegrationTest.assertProject(tempFolder, projectName, options.appid).wait();
 		});
+
+		describe("project name validation tests", () => {
+			let validProjectName = "valid";
+			let invalidProjectName = "1invalid";
+
+			beforeEach(() => {
+				hasPromptedForString = false;
+			})
```
Comment (rosen-vladimirov) [inline test/project-service.ts]: missing `;`
In fact this is my mistake, linter should catch such errors, but somehow I've excluded tests from linting :crying_cat_face:

**Comment by rosen-vladimirov** [comment]
:+1:  awesome work

**Pull request merged**